jep/401/README.adoc

```diff
@@ -0,0 +1,229 @@
+
+
+== Motivation
+
+Jenkins uses `pageHeader.jelly` file to specify the content of the header. Although it is valid, if a user wants to modify the header to perform some branding operation, Jenkins header branding capabilities are limited.
+
+There exist some plugins, like: https://plugins.jenkins.io/simple-theme-plugin/[simple-theme-plugin], that allow Jenkins users to customize some parts like CSS and/or Javascript. On the other hand, if a Jenkins user wants to customize/modify some additional business functionality on some menus and search bar, then there is no approach beyond updating/overriding `pageHeader.jelly` from the Jenkins core, which would be a problem on updating the instance due to conflicts. In addition, if the user wants to customize these behaviors it would be good to have those features as REST endpoints. So, this approach is about providing not only UI capabilities, it is also providing extra functionality and better integration.
+
+== Reasoning
+
+We will use an example of a Jenkins user that wants to update Jenkins header through an ad hoc Jenkins plugin that follows the principles provided on Specification section:
+
+=== Jenkins core
+
+* Let’s consider the following definition of the `Header` on: `core/src/main/java/jenkins/views/Header.java`
+
+```
+package jenkins.views;
+
+import hudson.ExtensionPoint;
+
+public abstract class Header implements ExtensionPoint {
+
+    /**
+     * Check if the header is enabled. By default it is if installed,
+     * but the logic is deferred in the plugins.
+     * @return
+     */
+    public abstract boolean isHeaderEnabled();
+
+}
+```
+
+* Let’s consider the following implementation of the Jenkins header on: `core/src/main/java/jenkins/views/JenkinsHeader.java`
+
+```
+package jenkins.views;
+
+import hudson.Extension;
+
+@Extension(ordinal = Integer.MIN_VALUE)
+public class JenkinsHeader extends Header {
+
+    @Override
+    public boolean isHeaderEnabled() {
+        return true;
+    }
+}
+```
+
+* As mentioned before, the `Functions` method `header()` will retrieve the available headers via `Extension` lookup that are enabled and for those obtained it will provide the one with max priority (ordinal value)
+
+```
+[...]
+@Restricted(NoExternalUse.class)
+@CheckForNull
+public static Header header() {
+    List<Header> headers = ExtensionList.lookup(Header.class).stream()
+                .filter(header -> header.isHeaderEnabled())
+                .collect(Collectors.toList());
+    if (headers.size() > 0) {
+        if (headers.size() > 1) {
+            LOGGER.warning("More than one configured header. This should not happen. Serving the Jenkins default header and please review");
+        } else {
+            return headers.get(0);
+        }
+    }
+    return new JenkinsHeader();
```

Comment:
> Suggestion: just return the one with higher priority

IIRC `ExtensionList.lookup` takes into account the optional value to order it in a transparent manner.

> we should provide something that is working easily for the developers, and thus, not requirement "advance" knowledge like the usage of an ExtensionFilter to ensure it's working as expected.

I see your point but at the same time we want to limit the number of active implementations to one. cc/ @fcojfernandez WDYT?

Reply:
> As JenkinsHeader seems to be part of the headers, if you provide any other implemention, you will get the warning + the JenkinsHeader. [...] I have not tested it manually, what was the result of your tests?

If you don't provide an ExtensionFilter, yes, it will show the WARNING on the log and render the default Jenkins one.